sql: TestCopyTransaction failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/29d1f23b0dba56a5fe58348993924722c5a48658

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestCopyTransaction PKG=github.com/cockroachdb/cockroach/pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=980473&tab=buildLog

```
I181023 05:46:58.745785 44080 storage/replica_command.go:298  [n1,split,s1,r20/1:/{Table/23-Max}] initiating a split of this range at key /Table/50 [r21]
I181023 05:46:58.767972 43441 server/testserver.go:446  had 20 ranges at startup, expected 22
I181023 05:46:58.814361 44100 storage/replica_command.go:298  [n1,split,s1,r21/1:/{Table/50-Max}] initiating a split of this range at key /Table/51 [r22]
I181023 05:46:59.212153 44178 sql/event_log.go:126  [n1,client=127.0.0.1:59640,user=root] Event: "create_database", target: 52, info: {DatabaseName:d Statement:CREATE DATABASE d User:root}
I181023 05:46:59.218603 44049 storage/replica_command.go:298  [n1,split,s1,r22/1:/{Table/51-Max}] initiating a split of this range at key /Table/52 [r23]
I181023 05:46:59.316378 44178 sql/event_log.go:126  [n1,client=127.0.0.1:59640,user=root] Event: "create_table", target: 53, info: {TableName:d.public.t Statement:CREATE TABLE t (i INT PRIMARY KEY) User:root}
I181023 05:46:59.358192 43711 storage/replica_proposal.go:211  [n1,s1,r23/1:/{Table/52-Max}] new range lease repl=(n1,s1):1 seq=3 start=1540273619.332712624,0 epo=1 pro=1540273619.336300786,0 following repl=(n1,s1):1 seq=2 start=1540273616.134597846,0 exp=1540273625.155179382,0 pro=1540273616.155269657,0
I181023 05:46:59.389957 44151 storage/replica_command.go:298  [n1,split,s1,r23/1:/{Table/52-Max}] initiating a split of this range at key /Table/53 [r24]
I181023 05:46:59.436892 43441 util/stop/stopper.go:537  quiescing; tasks left:
2      node.Node: batch
2      [async] kv.TxnCoordSender: heartbeat loop
1      storage.replicate: purgatory processing replica
1      storage.intentResolver: processing intents
1      [async] txnHeartbeat: aborting txn
1      [async] transport racer
1      [async] storage.split: processing replica
1      [async] closedts-rangefeed-subscriber
I181023 05:46:59.437857 43441 util/stop/stopper.go:537  quiescing; tasks left:
2      node.Node: batch
2      [async] kv.TxnCoordSender: heartbeat loop
1      storage.intentResolver: processing intents
1      [async] txnHeartbeat: aborting txn
1      [async] transport racer
1      [async] storage.split: processing replica
1      [async] closedts-rangefeed-subscriber
W181023 05:46:59.440010 44151 internal/client/txn.go:532  [n1,split,s1,r23/1:/{Table/52-Max}] failure aborting transaction: node unavailable; try another peer; abort caused by: log-range-event: failed to send RPC: sending to all 1 replicas failed; last error: (err: node unavailable; try another peer) <nil>
E181023 05:46:59.451731 44151 storage/queue.go:793  [n1,split,s1,r23/1:/{Table/52-Max}] unable to split [n1,s1,r23/1:/{Table/52-Max}] at key "/Table/53": split at key /Table/53 failed: log-range-event: failed to send RPC: sending to all 1 replicas failed; last error: (err: node unavailable; try another peer) <nil>
I181023 05:46:59.440471 43441 util/stop/stopper.go:537  quiescing; tasks left:
2      node.Node: batch
1      storage.intentResolver: processing intents
1      [async] txnHeartbeat: aborting txn
1      [async] transport racer
1      [async] storage.split: processing replica
1      [async] kv.TxnCoordSender: heartbeat loop
1      [async] closedts-rangefeed-subscriber
I181023 05:46:59.453827 43441 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] transport racer
1      [async] storage.split: processing replica
1      [async] kv.TxnCoordSender: heartbeat loop
1      [async] closedts-rangefeed-subscriber
I181023 05:46:59.454565 43441 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] transport racer
I181023 05:46:59.648018 43484 kv/transport_race.go:91  transport race promotion: ran 21 iterations on up to 914 requests
--- FAIL: TestCopyTransaction (3.80s)
	copy_in_test.go:419: pq: restart transaction: HandledRetryableTxnError: TransactionAbortedError(ABORT_REASON_TIMESTAMP_CACHE_REJECTED_POSSIBLE_REPLAY): "sql txn" id=0ab3f4be key=/Table/53/1/2/0 rw=true pri=0.01665531 iso=SERIALIZABLE stat=ABORTED epo=0 ts=1540273619.332712624,1 orig=1540273619.325646772,0 max=1540273619.332712624,0 wto=false rop=false seq=2

```

--- a/pkg/sql/copy_in_test.go
+++ b/pkg/sql/copy_in_test.go
@@ -16,7 +16,6 @@ package sql_test
 
 import (
 	"context"
-	gosql "database/sql"
 	"fmt"
 	"math"
 	"math/rand"
@@ -25,7 +24,6 @@ import (
 	"testing"
 	"time"
 
-	"github.com/cockroachdb/cockroach-go/crdb"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
 	"github.com/cockroachdb/cockroach/pkg/sql/tests"
@@ -64,6 +62,18 @@ func TestCopyNullInfNaN(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	txn, err := db.Begin()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	stmt, err := txn.Prepare(pq.CopyIn(
+		"t", "i", "f", "s", "b", "d", "t",
+		"ts", "n", "o", "e", "u", "ip", "tz"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
 	input := [][]interface{}{
 		{nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil},
 		{nil, math.Inf(1), nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil},
@@ -71,26 +81,22 @@ func TestCopyNullInfNaN(t *testing.T) {
 		{nil, math.NaN(), nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil},
 	}
 
-	if err := crdb.ExecuteTx(context.Background(), db, nil, func(txn *gosql.Tx) error {
-		stmt, err := txn.Prepare(pq.CopyIn(
-			"t", "i", "f", "s", "b", "d", "t",
-			"ts", "n", "o", "e", "u", "ip", "tz"))
+	for _, in := range input {
+		_, err = stmt.Exec(in...)
 		if err != nil {
-			return err
+			t.Fatal(err)
 		}
-
-		for _, in := range input {
-			_, err = stmt.Exec(in...)
-			if err != nil {
-				return err
-			}
-		}
-		_, err = stmt.Exec()
-		if err != nil {
-			return err
-		}
-		return stmt.Close()
-	}); err != nil {
+	}
+	_, err = stmt.Exec()
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = stmt.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = txn.Commit()
+	if err != nil {
 		t.Fatal(err)
 	}
 
@@ -152,6 +158,16 @@ func TestCopyRandom(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	txn, err := db.Begin()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	stmt, err := txn.Prepare(pq.CopyInSchema("d", "t", "id", "n", "o", "i", "f", "e", "t", "ts", "s", "b", "u", "ip", "tz"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
 	rng := rand.New(rand.NewSource(0))
 	types := []sqlbase.ColumnType_SemanticType{
 		sqlbase.ColumnType_INT,
@@ -168,37 +184,35 @@ func TestCopyRandom(t *testing.T) {
 		sqlbase.ColumnType_INET,
 		sqlbase.ColumnType_TIMESTAMPTZ,
 	}
+
 	var inputs [][]interface{}
 
-	if err := crdb.ExecuteTx(context.Background(), db, nil, func(txn *gosql.Tx) error {
-		stmt, err := txn.Prepare(pq.CopyInSchema("d", "t", "id", "n", "o", "i", "f", "e", "t", "ts", "s", "b", "u", "ip", "tz"))
+	for i := 0; i < 100; i++ {
+		row := make([]interface{}, len(types))
+		for j, t := range types {
+			var ds string
+			if j == 0 {
+				// Special handling for ID field
+				ds = strconv.Itoa(i)
+			} else {
+				d := sqlbase.RandDatum(rng, sqlbase.ColumnType{SemanticType: t}, false)
+				ds = tree.AsStringWithFlags(d, tree.FmtBareStrings)
+			}
+			row[j] = ds
+		}
+		_, err = stmt.Exec(row...)
 		if err != nil {
-			return err
+			t.Fatal(err)
 		}
+		inputs = append(inputs, row)
+	}
 
-		inputs = nil
-		for i := 0; i < 100; i++ {
-			row := make([]interface{}, len(types))
-			for j, t := range types {
-				var ds string
-				if j == 0 {
-					// Special handling for ID field
-					ds = strconv.Itoa(i)
-				} else {
-					d := sqlbase.RandDatum(rng, sqlbase.ColumnType{SemanticType: t}, false)
-					ds = tree.AsStringWithFlags(d, tree.FmtBareStrings)
-				}
-				row[j] = ds
-			}
-			_, err = stmt.Exec(row...)
-			if err != nil {
-				return err
-			}
-			inputs = append(inputs, row)
-		}
-
-		return stmt.Close()
-	}); err != nil {
+	err = stmt.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = txn.Commit()
+	if err != nil {
 		t.Fatal(err)
 	}
 
@@ -381,34 +395,37 @@ func TestCopyTransaction(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	if err := crdb.ExecuteTx(context.Background(), db, nil, func(txn *gosql.Tx) error {
-		// Note that, at least with lib/pq, this doesn't actually send a Parse msg
-		// (which we wouldn't support, as we don't support Copy-in in extended
-		// protocol mode). lib/pq has magic for recognizing a Copy.
-		stmt, err := txn.Prepare(pq.CopyIn("t", "i"))
-		if err != nil {
-			return err
-		}
+	txn, err := db.Begin()
+	if err != nil {
+		t.Fatal(err)
+	}
 
-		const val = 2
+	// Note that, at least with lib/pq, this doesn't actually send a Parse msg
+	// (which we wouldn't support, as we don't support Copy-in in extended
+	// protocol mode). lib/pq has magic for recognizing a Copy.
+	stmt, err := txn.Prepare(pq.CopyIn("t", "i"))
+	if err != nil {
+		t.Fatal(err)
+	}
 
-		_, err = stmt.Exec(val)
-		if err != nil {
-			return err
-		}
+	const val = 2
 
-		if err = stmt.Close(); err != nil {
-			return err
-		}
+	_, err = stmt.Exec(val)
+	if err != nil {
+		t.Fatal(err)
+	}
 
-		var i int
-		if err := txn.QueryRow("SELECT i FROM d.t").Scan(&i); err != nil {
-			return err
-		} else if i != val {
-			return fmt.Errorf("expected 1, got %d", i)
-		}
-		return nil
-	}); err != nil {
+	if err = stmt.Close(); err != nil {
+		t.Fatal(err)
+	}
+
+	var i int
+	if err := txn.QueryRow("SELECT i FROM d.t").Scan(&i); err != nil {
+		t.Fatal(err)
+	} else if i != val {
+		t.Fatalf("expected 1, got %d", i)
+	}
+	if err := txn.Commit(); err != nil {
 		t.Fatal(err)
 	}
 }

